[SIMD] Intel support for the remaining conversion opcodes

hyjorc1 opened the pull request:
#### a068c52c28a8dbb7a2e8a8a544856092657b2b06
<pre>
[SIMD] Intel support for the remaining conversion opcodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=249418">https://bugs.webkit.org/show_bug.cgi?id=249418</a>
rdar://103411760

Reviewed by Yusuke Suzuki.

Add support for conversion operations:
- i32x4.trunc_sat_f32x4_s(a: v128) -&gt; v128
- i32x4.trunc_sat_f32x4_u(a: v128) -&gt; v128
- f32x4.convert_i32x4_u(a: v128) -&gt; v128
<a href="https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#conversions">https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#conversions</a>

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorTruncSat):
(JSC::MacroAssemblerX86_64::vectorTruncSatUnsignedFloat32):
(JSC::MacroAssemblerX86_64::vectorConvertUnsigned):
(JSC::MacroAssemblerX86_64::vectorMulSat):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pblendw_i8rr):
(JSC::X86Assembler::vmaxps_rrr):
(JSC::X86Assembler::vmaxpd_rrr):
(JSC::X86Assembler::vminps_rrr):
(JSC::X86Assembler::vminpd_rrr):
(JSC::X86Assembler::vcmpunordps_rrr):
(JSC::X86Assembler::vcmpleps_rrr):
(JSC::X86Assembler::vcmpltps_rrr):
(JSC::X86Assembler::vcvttps2dq_rr):
(JSC::X86Assembler::vpand_rrr):
(JSC::X86Assembler::vpslld_i8rr):
(JSC::X86Assembler::pblendw_rr): Deleted.
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDV_V):
(JSC::Wasm::AirIRGenerator64::addSIMDRelOp):

Canonical link: <a href="https://commits.webkit.org/257965@main">https://commits.webkit.org/257965@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eddd82c05c830b7f18c0ff5ac47fbd6309c7c2f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100516 "22 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33564 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/109829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10592 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106295 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/91039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87081 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/868 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3392 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29141 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9509 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89964 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5453 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/5196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20114 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3038,15 +3038,65 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
     {
-        ASSERT(scalarTypeIsFloatingPoint(simdInfo.lane));
-        ASSERT(simdInfo.signMode != SIMDSignMode::None);
+        ASSERT(supportsAVXForSIMD());
```
We are about to replace `supportsAVXForSIMD` with `supportsAVX`, so let's use `supportsAVX`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3038,15 +3038,65 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        int32_t n = bitwise_cast<int32_t>(2147483647.0f);
+        move(TrustedImm32(n), tmp);
+        move32ToFloat(tmp, scratchFPR1);
+        vectorSplatFloat32(scratchFPR1, scratchFPR1);                       // scratchFPR1 = f[0x80000000]x4
```
We should make it load from memory with `static constexpr floats[] = { ... }`. Since emitting these instructions are (1) costly in terms of machine code size, and (2) move32ToFloat is not so cheap.
See the other existing code like `vectorTruncSatSignedFloat64`.

Constellation replied:
You can materialize an address in scratchGPR and use vsubps_mrr for example.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3038,15 +3038,65 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        m_assembler.vpxor_rrr(scratchFPR2, scratchFPR1, dest);                  // convert positive-overflow lane to 0x7FFFFFFF
+    }
+
+    void vectorTruncSatUnsignedFloat32(FPRegisterID src, FPRegisterID dest, RegisterID tmp, FPRegisterID scratchFPR1, FPRegisterID scratchFPR2)
```
Let's use `scratchGPR` name instead of `tmp`.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3254,9 +3304,17 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+    void vectorConvertUnsigned(FPRegisterID src, FPRegisterID dst, FPRegisterID scratch)
     {
-        UNUSED_PARAM(input); UNUSED_PARAM(dest); UNUSED_PARAM(scratch);
+        ASSERT(supportsAVXForSIMD());
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3038,15 +3038,65 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        //     1. saturate positive-overflow integer to 0xFFFFFFFF
+        //     2. saturate negative-overflow integer to 0
+        //     3. convert NaN or -0 to 0.    
+        
```
Add `ASSERT(supportsAVX());`

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4947,6 +4950,14 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/subps
+        // VEX.128.0F.WIG 5C /r VSUBPS xmm1,xmm2, xmm3/m128
+        // B    NA    ModRM:reg (w)    VEX.vvvv (r)    ModRM:r/m (r)    NA
+        m_formatter.vexNdsLigWigTwoByteOp(PRE_SSE_00, OP2_SUBPS_VpsWps, (RegisterID)dest, (RegisterID)left, (RegisterID)base, offset);
```
base is already `ReigsterID`, remove this `(RegisterID)` cast for base.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5222,6 +5253,18 @@ class X86Assembler {
+        if (condition == PackedCompareCondition::Equal || condition == PackedCompareCondition::NotEqual)
+            m_formatter.vexNdsLigWigCommutativeTwoByteOp(PRE_SSE_00, OP2_CMPPS_VpsWpsIb, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
+        else
+            m_formatter.vexNdsLigWigTwoByteOp(PRE_SSE_00, OP2_CMPPS_VpsWpsIb, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
```
Ditto, remove `(RegisterID)` cast to base.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -6997,6 +7065,14 @@ class X86Assembler {
+                std::swap(a, base);
+            vexNdsLigWigTwoByteOp(simdPrefix, opcode, dest, a, base, offset);
+        }
+
```
This is wrong since swapping `a` and `base` causes completely different meaning.
Commutative cannot be defined for instructions one side is memory and one side is register.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5222,6 +5253,18 @@ class X86Assembler {
+        // B    NA    ModRM:reg (w)    VEX.vvvv (r)    ModRM:r/m (r)    Imm8
+        if (condition == PackedCompareCondition::Equal || condition == PackedCompareCondition::NotEqual)
+            m_formatter.vexNdsLigWigCommutativeTwoByteOp(PRE_SSE_00, OP2_CMPPS_VpsWpsIb, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
+        else
```
So, remove this part. Commutative instructions cannot take memory. Always use `vexNdsLigWigTwoByteOp`

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[SIMD] Intel support for the remaining conversion opcodes
https://bugs.webkit.org/show_bug.cgi?id=249418
rdar://103411760

Reviewed by Yusuke Suzuki.

Add support for conversion operations:
- i32x4.trunc_sat_f32x4_s(a: v128) -> v128
- i32x4.trunc_sat_f32x4_u(a: v128) -> v128
- f32x4.convert_i32x4_u(a: v128) -> v128
https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#conversions

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorTruncSat):
(JSC::MacroAssemblerX86_64::vectorTruncSatUnsignedFloat32):
(JSC::MacroAssemblerX86_64::vectorConvertUnsigned):
(JSC::MacroAssemblerX86_64::vectorMulSat):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pblendw_i8rr):
(JSC::X86Assembler::vmaxps_rrr):
(JSC::X86Assembler::vmaxpd_rrr):
(JSC::X86Assembler::vminps_rrr):
(JSC::X86Assembler::vminpd_rrr):
(JSC::X86Assembler::vcmpunordps_rrr):
(JSC::X86Assembler::vcmpleps_rrr):
(JSC::X86Assembler::vcmpltps_rrr):
(JSC::X86Assembler::vcvttps2dq_rr):
(JSC::X86Assembler::vpand_rrr):
(JSC::X86Assembler::vpslld_i8rr):
(JSC::X86Assembler::pblendw_rr): Deleted.
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDV_V):
(JSC::Wasm::AirIRGenerator64::addSIMDRelOp):

Canonical link: https://commits.webkit.org/257965@main
